### ISSUE TITLE
Add some events in order too hook into

### ISSUE THREAD

**Issue by KeKs0r** [balderdashy/waterline#839]
Hi,

I am using waterline without sails and have sometimes to need to integrate better with waterline. In order to allow applications to hook better into waterline for very specific cases I would like to propose an event system that triggers events on:
- register Connection
- register Model
- migration
- Initialization

Currently I would need this, because I am using sails-postgres and added some views manually, but map them with waterline models. The problem is, that the views are not handled with the migration in my test case (=drop) and the existance of the view prevents waterline to delete the referenced table. 

But in general, I think the concept of events, would make waterline more flexible for such use cases, before e.g. a complete concept for views or other things are finished.

**Comment by devinivy** [comment]
I think this is a good idea for a feature!  In the meantime, I think this would be pretty easy to add into the `drop` method of the `sails-postgres` adapter.  Might be worth looking into.

**Comment by tjwebb** [comment]
I like this. Yea, currently sails orm hook fires these events, but there's a bunch of code involved that calls `waterline.initialize` etc. 

I think this is what would need to happen:
- add this as part of a future adapter spec
- add a "events" test to the waterline-adapter-spec
- implement in the adapters (I can do sails-postgresql and sails-sqlserver)

A few moving parts, but I think not a ton of work. @devinivy would you be willing to help out with this?

@particlebanana what are your thoughts?

**Comment by devinivy** [comment]
I would be willing to help!  I'm familiar with disk, memory, dynamodb, and mysql adapters.  A great place to start would be to have a test to work from.

**Comment by KeKs0r** [comment]
(Short Off topic)
For my actual scenario with the view I have done this in my model definition:
```Javascript
module.exports = {
    tableName: 'user_rankings_global',
    attributes: {
        userid: {
            type: 'string'
        },
        nickname: {
            type: 'string'
        },
        points: {
            type: 'integer'
        }
    },
    // This overwrites normal migration strategy
    sync: function(cb){
        if(this.migrate === 'drop'){
            var query = 'DROP VIEW IF EXISTS ' + this.tableName;
            this.query(query,cb);
        }else{
            cb();
        }
    }
};
```
Also this approach solves the issue: https://github.com/balderdashy/sails-postgresql/pull/135

----- Back to the event system -----
I think the events should be implemented solely on waterline and not on the adapters, but maybe not every event can be. I will look into the internals a little bit in order to see if this can be done in waterline.

For me there are two more questions:
1. Where do we want to trigger events? (I already provided some ideas, any additional input on that?)
2. How do we trigger the events?
Should the execution wait for all eventlistener to finish? I think this should be the case, because if I trigger an event within "initialize" that changes some parts I only want to initialization function to continue after my listeners are processed. In order to achieve that, I am currently using this library in a project of mine: https://www.npmjs.com/package/async-eventemitter
The same logic could be achieved with async, but I like the API of the async event emitter.

**Comment by devinivy** [comment]
@dmarcelino @particlebanana @tjwebb could this be more for 0.11 or are we booked?

**Comment by dmarcelino** [comment]
I actually need this on `sails-orientdb` to detect when all collections have been created so I can create links (a sort of association in orientdb world) for 1-1 and 1-N between tables. Currently I'm using a hack where I monitor when `describe` and `define` have been called and try to infer if all tables are created (which is required in order to create such links). 

:+1:  for `0.11`

**Comment by dmarcelino** [comment]
> implement in the adapters (I can do sails-postgresql and sails-sqlserver)

I'm actually of the opinion waterline should be triggering these events. I mean, all adapter methods (with the exception of custom methods) are called by waterline itself, so waterline *knows* when a connection is registered or a table is created, etc. We can do it once in waterline instead of repeating it across all adapters, unless I'm missing something...

**Comment by dmarcelino** [comment]
One issue that could benefit from said events/hooks is: Feature: implement caching mechanism #200.

**Comment by sailsbot** [comment]
Thanks for posting, @KeKs0r.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by dmarcelino** [comment]
Given that @KeKs0r, @devinivy, @tjwebb and me like this feature request how about we give it a second thought, @sailsbot?

**Comment by sailsbot** [comment]
Thanks for posting, @KeKs0r.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
